File: ts/src/safesocket.ts
import { execFile } from "node:child_process";
import { readFile, readlink } from "node:fs/promises";
import { platform } from "node:os";
import { join } from "node:path";
import { promisify } from "node:util";

export const LOCAL_API_HOST = "local-tailscaled.sock";
export const CURRENT_CAP_VERSION = 131;

export interface PortAndToken {
  port: number;
  token: string;
}

/** Return the default socket path for the current platform. */
export function defaultSocketPath(): string {
  if (platform() === "darwin") {
    return "/var/run/tailscaled.socket";
  }
  // Linux and other Unix
  return "/var/run/tailscale/tailscaled.sock";
}

/** Attempt to discover macOS TCP port and token for tailscaled. */
export async function localTcpPortAndToken(): Promise<PortAndToken | undefined> {
  if (platform() !== "darwin") {
    return undefined;
  }

  // Try lsof method first (macOS GUI app)
  const result = await readMacosSameUserProof();
  if (result) return result;

  // Try filesystem method (macOS system extension)
  return readMacsysSameUserProof();
}

const execFileP = promisify(execFile);

async function readMacosSameUserProof(): Promise<PortAndToken | undefined> {
  try {
    const uid = process.getuid?.();
    if (uid === undefined) return undefined;

    const { stdout: output } = await execFileP("lsof", [
      "-n",
      "-a",
      `-u${uid}`,
      "-c",
      "IPNExtension",
      "-F",
    ]);
    return parseLsofOutput(output);
  } catch {
    return undefined;
  }
}

/** Parse lsof -F output looking for sameuserproof-PORT-TOKEN. */
export function parseLsofOutput(output: string): PortAndToken | undefined {
  const needle = ".tailscale.ipn.macos/sameuserproof-";
  for (const line of output.split("\n")) {
    const idx = line.indexOf(needle);
    if (idx === -1) continue;
    const rest = line.slice(idx + needle.length);
    const dash = rest.indexOf("-");
    if (dash === -1) continue;
    const portStr = rest.slice(0, dash);
    const token = rest.slice(dash + 1);
    const port = parseInt(portStr, 10);
    if (!isNaN(port)) {
      return { port, token };
    }
  }
  return undefined;
}

async function readMacsysSameUserProof(
  sharedDir = "/Library/Tailscale",
): Promise<PortAndToken | undefined> {
  try {
    const portPath = join(sharedDir, "ipnport");
    const portStr = await readlink(portPath, "utf-8");
    const port = parseInt(portStr, 10);
    if (isNaN(port)) return undefined;
    const tokenPath = join(sharedDir, `sameuserproof-${port}`);
    const tokenRaw = await readFile(tokenPath, "utf-8");
    const token = tokenRaw.trim();
    return { port, token };
  } catch {
    return undefined;
  }
}

File: python/src/tslocal/_safesocket.py
"""Platform-specific socket connection to tailscaled."""

from __future__ import annotations

import os
import platform
import subprocess
import sys
from pathlib import Path
from typing import NamedTuple


LOCAL_API_HOST = "local-tailscaled.sock"
CURRENT_CAP_VERSION = 138


class PortAndToken(NamedTuple):
    port: int
    token: str


def default_socket_path() -> str:
    """Return the default socket path for the current platform."""
    if platform.system() == "Darwin":
        return "/var/run/tailscaled.socket"
    # Linux and other Unix
    return "/var/run/tailscale/tailscaled.sock"


def local_tcp_port_and_token() -> PortAndToken | None:
    """Attempt to discover macOS TCP port and token for tailscaled.

    Returns None if not on macOS or discovery fails.
    """
    if platform.system() != "Darwin":
        return None

    # Try lsof method first (macOS GUI app)
    result = _read_macos_same_user_proof()
    if result is not None:
        return result

    # Try filesystem method (macOS system extension)
    return _read_macsys_same_user_proof()


def _read_macos_same_user_proof() -> PortAndToken | None:
    """Try lsof to find IPNExtension sameuserproof."""
    try:
        uid = os.getuid()
        result = subprocess.run(
            ["lsof", "-n", "-a", f"-u{uid}", "-c", "IPNExtension", "-F"],
            capture_output=True,
            text=True,
            timeout=5,
        )
        return parse_lsof_output(result.stdout)
    except (OSError, subprocess.TimeoutExpired):
        return None


def parse_lsof_output(output: str) -> PortAndToken | None:
    """Parse lsof -F output looking for sameuserproof-PORT-TOKEN."""
    needle = ".tailscale.ipn.macos/sameuserproof-"
    for line in output.splitlines():
        idx = line.find(needle)
        if idx == -1:
            continue
        rest = line[idx + len(needle) :]
        dash = rest.find("-")
        if dash == -1:
            continue
        port_str = rest[:dash]
        token = rest[dash + 1 :]
        try:
            port = int(port_str)
            return PortAndToken(port, token)
        except ValueError:
            continue
    return None


def _read_macsys_same_user_proof(
    shared_dir: str = "/Library/Tailscale",
) -> PortAndToken | None:
    """Try filesystem at /Library/Tailscale/ for system extension."""
    try:
        port_path = Path(shared_dir) / "ipnport"
        port_str = os.readlink(port_path)
        port = int(port_str)
        token_path = Path(shared_dir) / f"sameuserproof-{port}"
        token = token_path.read_text().strip()
        return PortAndToken(port, token)
    except (OSError, ValueError):
        return None
